addons/io/org.openhab.io.hueemulation/src/main/java/org/openhab/io/hueemulation/internal/HueEmulationServlet.java

```diff
@@ -534,44 +486,86 @@ private HueDevice itemToDevice(Item item) {
         }
 
-        HueDevice d = new HueDevice(hueState, item.getLabel(), item.getName());
+        HueDevice d = new HueDevice(hueState, item.getLabel(), key);
         return d;
     }
 
     /**
-     * Gets all items that match our tag
-     *
-     * @return
+     * Gets and syncs all items tagged for voice.
      */
-    private Collection<Item> getTaggedItems() {
-        Collection<Item> items = new LinkedList<Item>();
+    private synchronized TreeMap<Integer, HueDevice> getHueDevices() {
+        TreeMap<Integer, HueDevice> returnMap = new TreeMap<Integer, HueDevice>();
+        HashMap<String, Item> taggedItems = new HashMap<String, Item>();
+
+        // if we modify our internal map, persist it to disk
+        boolean modified = false;
+
+        // get all tagged items
         for (Item item : itemRegistry.getItems()) {
             for (String tag : item.getTags()) {
                 if (ArrayUtils.contains(SUPPORTED_TAGS, tag)) {
-                    items.add(item);
+                    taggedItems.put(item.getName(), item);
+                    if (!deviceMap.containsValue(item.getName())) {
+                        // hue devices are assigned a numeric number starting with 1, if a device is
+                        // removed that number is not used again. Not sure how high this id can get
+                        // not worrying about it here
+                        Integer next = deviceMap.size() == 0 ? 1 : new Integer(deviceMap.lastKey().intValue() + 1);
+                        deviceMap.put(next, item.getName());
+                        modified = true;
+                    }
                     break;
                 }
             }
         }
-        return items;
+
+        Set<Integer> keysToRemove = new HashSet<>();
+
+        // clean up removed entries
+        for (Map.Entry<Integer, String> entry : deviceMap.entrySet()) {
+            String itemName = entry.getValue();
+            if (!taggedItems.containsKey(itemName)) {
+                keysToRemove.add(entry.getKey());
+                modified = true;
+            }
+        }
+
+        for (Integer key : keysToRemove) {
+            deviceMap.remove(key);
+        }
+
+        // for each entry, lookup the item and convert it to a hue device
+        for (Integer key : deviceMap.keySet()) {
+            try {
+                returnMap.put(key, itemToDevice(itemRegistry.getItem(deviceMap.get(key)), key));
+            } catch (ItemNotFoundException e) {
+                logger.warn("Could not find item", e);
+            }
+        }
+
+        if (modified) {
+            JsonWriter writer = null;
+            try {
+                writer = new JsonWriter(new FileWriter(ITEM_FILE));
+                gson.toJson(deviceMap, new TypeToken<Map<Integer, String>>() {
+                }.getType(), writer);
+            } catch (IOException e) {
+                logger.error("Could not persist item cache", e);
```

Comment:
This is a separate situation: It tries to write a file to disk and fails - so the user needs to be notified as he has some fatal issue in his system.